Points & Pixels as Page Layout Units

sidewayss opened the pull request:
MuseScore Feature Request: 106861
here: https://musescore.org/en/node/106861

This PR replaces #2573.

lasconic commented on libmscore/page.cpp:
```diff
@@ -417,36 +441,34 @@ void PageFormat::read(XmlReader& e, Score* score)
-      xml.etag();
-
-      xml.etag();
+    xml.stag("page-layout");
```
Sorry... the indentation is wrong here.

sidewayss commented:
I fixed the indentation.

lasconic commented on mscore/exportxml.cpp:
```diff
@@ -1156,12 +1156,12 @@ void ExportMusicXml::credits(Xml& xml)
-      const double rm = getTenthsFromInches(pf->oddRightMargin());
-      //const double tm = getTenthsFromInches(pf->oddTopMargin());
-      const double bm = getTenthsFromInches(pf->oddBottomMargin());
+      const double h  = pf->size().height()   * PageFormat::SCALE_XML;
```
I don't get this. I think it's wrong. MusicXML needs all these values in 10th of space. Testing further.

sidewayss replied:
They are at 1/10th of a space.  Staff space == 5 pt/px.  1/10th staff space = 0.5pt/px.  PtPx = 72dpi.  SCALE_XML = 144dpi.
Is this possibly a problem is Staff space != 5pt/px?  Maybe that is a real problem.  I need to calculate SCALE_XML as a variable instead of making it a constant?  Or is this supposed to be relative to Staff Space?

lasconic replied:
I did a diff between a MusicXML file before and after this PR. If we exclude rounding it looks correct except for margins, which are inversed and some default-x values which are wrong, probably for the same reason.
```
<         <right-margin>56.6929</right-margin>
---
>         <right-margin>-55.1181</right-margin>
35c35
<         <right-margin>56.6929</right-margin>
---
>         <right-margin>-55.1181</right-margin>
```

sidewayss replied:
So are the MusicXML units fixed at 144dpi, or are they variable depending on the spatium?  It seems like your margin issue is unrelated to SCALE_XML.

lasconic commented:
And if there is one more thing I'm not sure about this PR, it's whether we want to save the unit in the score and so makes it a "score property" or if we don't save it and keep it as a "MuseScore preference".

lasconic commented:
Also, it works for my example (more or less, see above) but I must say I don't get 
```
They are at 1/10th of a space. Staff space = 5 pt/px. 1/10th staff space = 0.5pt/px. PtPx = 72dpi. SCALE_XML = 144dpi.
```

The staff space can be set by the user in mm/inch. So it's variable and defined at the top of any musicxml or mscx files. In MusicXML with the scaling element http://usermanuals.musicxml.com/MusicXML/MusicXML.htm#CT-MusicXML-scaling.htm and in MSCX with the Spatium element. SCALE_XML = 2, and not 144dpi...

lasconic commented:
The inversed right margin happens only with the My_First_Score loaded at the startup. This score doesn't have any page size information, and the size is set according to the user locale.

lasconic commented on mscore/musescore.cpp:
```diff
@@ -5048,11 +5048,11 @@ int main(int argc, char* av[])
             PageFormat pf;
             pf.setSize(&ps);
-            pf.setPrintableWidth(psf.width() - 20.0 / INCH);
+            pf.setPrintableWidth(psf.width() - 20.0 / MMPI);
```
This should be `pf.setPrintableWidth(psf.width() - 20.0 / DPMM);`

sidewayss replied:
That is clearly a typo on my part. Thanks for catching it.  I'll post these fixes in a few  minutes.

lasconic commented on libmscore/page.cpp:
```diff
@@ -288,15 +296,20 @@ void Page::scanElements(void* data, void (*func)(void*, Element*), bool all)
-      _oddTopMargin     = 10.0 / INCH;
-      _oddBottomMargin  = 20.0 / INCH;
-      _twosided         = true;
+    const int   marginMM  = 10;              // Default margin in millimeters
```
Here as well the indentation is changed.

lasconic commented on libmscore/page.cpp:
```diff
@@ -417,33 +441,31 @@ void PageFormat::read(XmlReader& e, Score* score)
-            xml.etag();
-            type = "odd";
-            }
+    xml.stag("page-layout");
```
indentation is wrong

sidewayss commented:
<code>The staff space can be set by the user in mm/inch. So it's variable and defined at the top of any musicxml or mscx files. In MusicXML with the scaling element http://usermanuals.musicxml.com/MusicXML/MusicXML.htm#CT-MusicXML-scaling.htm and in MSCX with the Spatium element. 
SCALE_XML = 2, and not 144dpi...</code>

SCALE_XML is a constant that equals 144.  So it will never equal 2.  The scaling seems to work OK with this as constant.  If you make it variable via the spatium, then exported musicxml files will all be exported as if the spatium was set to 5pt/px.  In other words, if you make SCALE_XML a variable based on the spatium, the spatium value is effectively ignored.

The MusicXML documentation you link to is unclear about this, because it does not acknowledge that staff space is variable, it only mentions a default value of 5.  I think part of the confusion here is that MuseScore gives the Staff Space value a double meaning:
1) original meaning, the distance between staff lines in a 5-line staff
2) a scale factor for the entire score, based specifically on the relationship to the default value of 5 pt/px.

I don't see MusicXML docs that show the staff space being used as a global scaling factor.  It seems to be a MuseScore addition to the meaning of Staff Space, not a MusicXML feature.  Is that correct?

sidewayss commented:
indentation and DPMM replacement fixed.

lasconic commented:
*SCALE_XML is a constant that equals 144. Really, check this :)*
https://github.com/musescore/MuseScore/pull/2624/files#diff-561aacb94c4695f64760cd81b78b17a6R98

About staff space or spatium, it only has a single meaning. It's the space between two lines of a staff. In MuseScore, a lot of thing are measure in this unit. In MusicXML, they prefer to use 1/10 of it.
This staff space has no reality until you associate it with an actual unit. MuseScore does it by saying one spatium = X mm (or inches). See Layout > Page settings. MusicXML does it with scaling elements which says more or less the same: a number of 1/10 of spatium = a number of millimeters. I still fail to see what 5 pt/px is :( but your code works on my example, if we ignore rounding errors.

`One spatium = Xmm = X * DPMM = X * 72 / 25.4`
**The default value of the spatium is `X = 1.764`**
And then indeed `1.764 * (72/25.4)` happens to do `1 spatium = 5.000px`!! I got it !

So if you assumer that 1 spatium = 5px in your code, then that will not work. If the user sets 1 spatium = 3mm in Layout > Page settings, then 1 spatium = 8.5039... and not 5.0...

sidewayss commented:
Yes, sorry.  SCALE_XML causes the scale to switch from 72dpi to 144dpi.  My misspeaking - I think of it all as 144.
Here is the strange thing: I just created two files where the Staff Space is set to 5 times the default, 8.819mm.  I exported using the current master code, then again using this branch's code.  I've attached the .mscz with the original scaling (1.764mm), and 3 exports.  The one with master5 in the name is the export using the master code using the 8.819mm staff space value.  When I open it in MuseScore it looks horrible.  The other two files are this branch's 8.819mm export (with the 5 in the file name) and it's 1.764mm export (the shortest .xml file name)
[Scores.zip](https://github.com/musescore/MuseScore/files/282829/Scores.zip)

So it looks to me like I am fixing a bug here, not breaking anything.  In the two "...5.xml" files search for the scaling tag.  It will include the millimeter and tenths values, which are identical across these two files.  But the values inside each individual visual element are constant in this branch's file, and they are multiplied by 5 in the master5.xml file.

Please confirm this or show me what I'm doing wrong.  Thanks.

sidewayss commented:
I have been playing with this some more. The first thing you have to remember is that this PR only affects PageFormat values.  It in no way affects the visual element values.  The page format values are 100% unaffected by the Staff Space parameter.  So staff spacing should not be a concern here.

This means that changing the millimeters formula in exportxml.cpp does not allow me to remove the SCALE_XML constant.  The MusicXML scaling element, and the millimeters and tenths class variables, have nothing to do with PageFormat attributes.  The page size doesn't get scaled, the margins don't get scaled, etc.

In playing with the tenths and millimeters variables I have seen that it really doesn't matter what tenths is set to, because millimeters scales appropriately via the formula.  tenths can be set to any positive value at all, and there is no difference in the rendering of the resulting export.  This is because the formula for millimeters is what matters, because it includes the spatium. So setting tenths to 1 is a cleaner value than 40, but it really doesn't matter what it is.  But if you set tenths to 1, then millimeters = 0.1764, the staff space value (in millimeters) divided by 10.  Thus the name "tenths".

In that sense some simplification could be achieved, in that tenths and millimeters might not need to exist as class variables, because they are only used in one place.  This block of code is the only place where defaults() is called, and defaults() is the only place where tenths and millimeters are used:
https://github.com/sidewayss/MuseScore/blob/e2d3f6be34703b22e9d3079c672d4fb3c2627920/mscore/exportxml.cpp#L4968-L4971
There is one other place where they are used, but that formula can be modified to eliminate the need for them, here:
https://github.com/sidewayss/MuseScore/blob/e2d3f6be34703b22e9d3079c672d4fb3c2627920/mscore/exportxml.cpp#L5016
This:
<code>
 / DPMM / millimeters * tenths
</code>
 could change to this:
<code>
 / (_score->spatium() * DPMM * PageFormat::SCALE_XML) / 10
</code>
and then no more class variables the values for tenths and millimeters are calculated the only place they are used, in defaults().  The other formula does get a bit more complex, I have to admit.  But I still like the idea of at least setting tenths to 1 instead of 40.  It truly illustrates the scaling in millimeters inside the xml file, and certainly eliminates the need for the tenths class variable.

sidewayss commented:
Travis build just failed and I don't understand the problem.  It's here:
diff -u part-breath-add.mscx
and here:
diff -u part-breath-uadd.mscx
The main change I made was to rebase to the latest master.  Is this something I need to fix?  Any advice would be helpful.

sidewayss commented:
I also have a concern about this block of code, which is currently disabled:
https://github.com/sidewayss/MuseScore/blob/e2d3f6be34703b22e9d3079c672d4fb3c2627920/mscore/exportxml.cpp#L5017-L5138
What is the plan for this code? Should I be updating it too? Will it be returning any time (soon)?

sidewayss commented:
[scaling.zip](https://github.com/musescore/MuseScore/files/284653/scaling.zip)

This has nothing to do with PageFormat values, it is about everything else that gets exported and converted.  The current code is creating a scaling factor, then applying the inverse of that scaling factor to all the notes, etc.  100% unnecessary.  getTenthsFromDots() can be replaced by * SCALE_XML and it works the same as with getTenthsFromDots().  It doesn't really make sense that it works in the current code.  I leave the millimeters and tenths values the same, stop scaling the notes, and it looks the same.

Check out the two .xml files in the attached zip.  This illustrates how it makes no sense.  They both have Staff Space set to 5 times the default value (8,819mm, 25pt/px).  The difference is that the code in the "new" file replaces getTenthsFromDots() with * SCALE_XML - so it ignores the Staff Space (spatium) scaling x5 when exporting those values.
The two files render IDENTICALLY in MuseScore.  Yet if you diff them you'll see that there are differences in measure width and other properties inside the xml.  How can they render identically?  Is there a bug in the MuseScore importxml files?

Is there another MusicXML rendering program I should use as a comparison?

Plus here is a bug in the current master conversions (lm is already converted above, it's converting twice, see the first line vs. the second line here, lm is outside the conversion in the first line and inside it in the second line):

sidewayss commented:
I have posted a branch that does these non-PageFormat scaling changes here:
https://github.com/sidewayss/MuseScore/tree/PtPx4a
In case you want to test them or see the code changes in exportxml.cpp
Really, SCALE_XML becomes more than a PageFormat option, and should probably be moved somewhere more generic, but that's a relatively minor concern.  Until I understand why those two files render the same, I can't say that this is completely correct.

lasconic commented:
I'm on something else right now but if by render the same, you mean once imported in MuseScore they look the same, it's possible that importxml ignores default-x, measure width etc...

sidewayss commented:
That is exactly what I mean.  When I open the files in MuseScore or soundslice they look exactly the same.
It can't ignore default-x or <measure width=xxx>.  That's part of what doesn't make sense.  If I change the measure width in the .mscz file then re-export, the values change, and the viewers see that change.  But somehow it's ignoring the relative scaling of these values, which continues to make no sense to me.  Maybe I should try this with a tenths value other than 1...

lasconic commented:
On import the measure width are ignored by MuseScore, probably by most of the musicxml importers out there.

sidewayss commented:
If, only in exportxml.cpp, I eliminate all the conversions, and I mean ALL of them, the file renders the same way.  The uses of SCALE_XML is page.cpp are necessary for the reading/writing of the mscx/z files (<i>or are they?</i>).  But exportxml.cpp only needs to set the tenths and millimeters values in the scaling element, the page format attributes and measures and notes are fine without any scaling.
So what do you want to do about this?  I think I should include these changes in this PR.  It would simplify exportxml.cpp dramatically.

sidewayss commented:
correction: I must keep 1 instance of SCALE_XML in exportxml.cpp, for the PageFormat values.  The PtPx4a branch is updated with this no-scaling of non-pageformat elements:
https://github.com/sidewayss/MuseScore/tree/PtPx4a
Attached is a zip containing an mscz and it's exported .xml file:
[NoScaling.zip](https://github.com/musescore/MuseScore/files/284918/NoScaling.zip)

sidewayss commented:
As for page.cpp.  I believe that the scaling from 72dpi to 144dpi is unnecessary, but changing it would break compatibility with previously saved files.  backwards compatibility could be handled with a version number check on the .mscx/z file.  It would only impact PageFormat::read(), which has 3 conversions.  Or it stays the way it is, and that's just part of the .mscx specification: 144dpi.
MusicXML is not .mscx, so this code in page.cpp has nothing to do with exportxml.cpp, and is only relevant to PageFormat values.

lasconic commented:
Ok, earlier I mentioned that I was not a fan of adding unit in the file format. I talked it through with @wschweer  and he doesn't like it either. The unit should be a property of MuseScore, probably set to inch/mm depending on the user locale by default and eventually changed to pixel if the user wants but there is no reason to store it in the file. @sidewayss can you remove this?

sidewayss commented:
I can remove the storage of the units easily.  But think about this for a moment:
MuseScore already changes the units based on the paper size chosen.  If I live in France, and I save a score with Letter size paper, which is in inches, what happens when I open that score later?  Does it default to mm, even though the paper size is Letter, and is specified in inches?

MuseScore is saving the Paper Size, though it is not switching the units when you select Paper Size.  To me that would be the better change: when you select A4 the units switch to mm, when you select Letter the units switch to inch, etc.  And then the units are stored with the score.  If a score has A4 paper size, it has mm units, if it has Letter paper size, it has inch units, it it's custom, then the user selects the units.

For me, I work on different types of scores.  Some of them are designed to be printed, and living in Latin America, Letter size is the paper most available to me.  But many of my new scores are specifically designed for the web, and are in pixels.  When I want to adjust the page size or margins on my web scores, which are all Custom paper size, I don't want to have to select Points/Pixels every time I open the Page Layout dialog.

I can revert this, but I added it for a reason: It was a hassle every time I changed page size or margins on my Custom page size scores.  If you select Letter paper size, do you really want to see the units in millimeters?  Doesn't anyone besides me work with different paper sizes that require different units?

sidewayss commented:
Some precedent to back up my position on storing the page units:
The other desktop publishing programs that let you specify the units in the Document Properties do it.  I have tested two popular open-source programs, Scribus and Inkscape, and they both store the units.
MS Excel and Open Office writer do not store the units, but they don't let you specify them either.  In the programs that do let you specify the units for the document (per document, not global), the units are also stored.

I rest my case.

sidewayss commented:
Here is a suggestion that might quell your fears about "storing the units":
1) In the Page Settings dialog convert the existing Units radio buttons to a single drop-down list.  This is not essential, but I think it tones down the importance of this option, for the users who don't want to concern themselves with it.

2) Add a 4th item to the list: "default: mm", where "mm" might be "in" depending on your locale.

If the user doesn't change the default units specified by their locale, then the units aren't stored.  You could store a "default" value, but not storing it is in line with backwards compatibility - either way. Qt has a QLocale class, which makes this easier.

This solves the two concerns that I can think of regarding "storing the units":
a) Sharing a file with someone overseas.  It might be convenient for that overseas user to see their own units, not the composer's.
b) MuseScore templates, which should open using the default, and which should be maintained by editing in MuseScore, not tampering inside the text file, if at all possible, right?

I'm willing to add to this PR in order to advance it.  You would be getting not only the functionality you desire, default values preserved across locales, but also a better default value based on locale (current default is hardcoded to "mm").  And yes, eventually it would be good to be able to override the locale-specified default with one the user can select in Preferences somewhere.  But one step at a time...

NOTE: QLocale has 3 variations on Imperial units, but fortunately the Inch is the same in all of them.

sidewayss commented:
I realize that there have been other priorities for the MuseScore crew over the summer, but I'm wondering when this PR might come up for consideration again.  If I have to, of course I will remove the saving of the units, but it really is the correct thing to do.

Jojo-Schmitz commented:
Well, it needs a rebase at least

sidewayss commented:
Absolutely.  I have a new branch rebased as of last night.  It would mean a new PR or a copying of those files to this branch.  I rebase it semi-regularly anyway just to keep up with bug fixes.  Though the latest master has some weird lyrics spacing :-)
